src/common/libflux/message.c

```diff
@@ -427,9 +384,9 @@ static int msg_append_route (flux_msg_t *msg,
     return 0;
 }
 
-static int iovec_to_msg (flux_msg_t *msg,
-                         struct msg_iovec *iov,
-                         int iovcnt)
+int flux_iovec_to_msg (flux_msg_t *msg,
```

Comment:
Why do these need to have the `flux_` prefix?  This will make them available publicly.

Reply:
This was done b/c `libflux-core.so` doesn't export any `flux_` prefixed functions, so the iovec functions weren't available in some places.  But now that I think about it, any code that needs the iovec functions could ldadd `src/common/libflux/libflux.la` instead of `src/common/libflx-core.la`.  Let me see how that works out.

Reply:
Well, this experiment didn't go as well as I thought it would.  There's so many symbols to resolve, and a nice chunk of "what to ldadd first" to ensure dependencies get resolved accordingly.

```
diff --git a/src/common/librouter/Makefile.am b/src/common/librouter/Makefile.am
index 756f5bb..4dcdd41 100644
--- a/src/common/librouter/Makefile.am
+++ b/src/common/librouter/Makefile.am
@@ -63,9 +63,15 @@ test_ldadd = \
         $(top_builddir)/src/common/librouter/librouter.la \
         $(top_builddir)/src/common/libtestutil/libtestutil.la \
         $(top_builddir)/src/common/libflux-internal.la \
-        $(top_builddir)/src/common/libflux-core.la \
-        $(top_builddir)/src/common/libtap/libtap.la \
+        $(top_builddir)/src/common/libflux/libflux.la \
+        $(top_builddir)/src/common/liblsd/liblsd.la \
+        $(top_builddir)/src/common/libev/libev.la \
+        $(top_builddir)/src/common/libutil/libutil.la \
+        $(top_builddir)/src/common/libccan/libccan.la \
+        $(top_builddir)/src/common/libtomlc99/libtomlc99.la \
         $(top_builddir)/src/common/libzmqutil/libzmqutil.la \
+        $(top_builddir)/src/common/libczmqcontainers/libczmqcontainers.la \
+        $(top_builddir)/src/common/libtap/libtap.la \
        $(ZMQ_LIBS)
 
 test_cppflags = \
```

worse yet I'm hitting segfaults in sharness tests.  I haven't been able to figure it out yet, although the backtrace of the segfault points to an unknown function.  A symbol collision wouldn't surprise me given what's going on above.

Hmmm.  Let me mull over this tonight and maybe something obviously simpler can come to mind.

Reply:
i can't think of an obviously clean solution here.  We don't want to move iovec into yet another convenience lib.  Would it be worthwhile to prefix the functions to something to indicate they are private and make sure they are accessible in the `libflux-core.map` file?  Like prefix them with `fprivate_` or something instead?

Although unintended, we get away with this on the reactor side b/c we static inline those two simple functions.  We'd have that problem there too if the functions were not inlined.

Reply:
Can we just add `libflux/libflux.la` right after `libflux-core.la` in the LDADD rules?  I didn't try to work through any tests, but the following patch resulted in a working broker and shmem connector for me:
```diff
diff --git a/src/broker/Makefile.am b/src/broker/Makefile.am
index f25cd2eaf..3c02fd731 100644
--- a/src/broker/Makefile.am
+++ b/src/broker/Makefile.am
@@ -67,6 +67,7 @@ flux_broker_LDADD = \
        $(builddir)/libbroker.la \
        $(top_builddir)/src/common/libflux-core.la \
        $(top_builddir)/src/common/libzmqutil/libzmqutil.la \
+       $(top_builddir)/src/common/libflux/libflux.la \
        $(top_builddir)/src/common/libpmi/libpmi_client.la \
        $(top_builddir)/src/common/libflux-internal.la \
        $(top_builddir)/src/common/libflux-optparse.la \
diff --git a/src/common/libflux/message.c b/src/common/libflux/message.c
index 51e402d2e..c6f0db869 100644
--- a/src/common/libflux/message.c
+++ b/src/common/libflux/message.c
@@ -384,7 +384,7 @@ static int msg_append_route (flux_msg_t *msg,
     return 0;
 }
 
-int flux_iovec_to_msg (flux_msg_t *msg,
+int iovec_to_msg (flux_msg_t *msg,
                        struct msg_iovec *iov,
                        int iovcnt)
 {
@@ -507,7 +507,7 @@ flux_msg_t *flux_msg_decode (const void *buf, size_t size)
         iovcnt++;
         p += n;
     }
-    if (flux_iovec_to_msg (msg, iov, iovcnt) < 0)
+    if (iovec_to_msg (msg, iov, iovcnt) < 0)
         goto error;
     free (iov);
     return msg;
@@ -1622,7 +1622,7 @@ void flux_msg_fprint (FILE *f, const flux_msg_t *msg)
     flux_msg_fprint_ts (f, msg, -1);
 }
 
-int flux_msg_to_iovec (const flux_msg_t *msg,
+int msg_to_iovec (const flux_msg_t *msg,
                        uint8_t *proto,
                        int proto_len,
                        struct msg_iovec **iovp,
diff --git a/src/common/libflux/message_private.h b/src/common/libflux/message_private.h
index 7923b8c10..8872d3082 100644
--- a/src/common/libflux/message_private.h
+++ b/src/common/libflux/message_private.h
@@ -61,11 +61,11 @@ struct msg_iovec {
     void *transport_data;
 };
 
-int flux_iovec_to_msg (flux_msg_t *msg,
+int iovec_to_msg (flux_msg_t *msg,
                        struct msg_iovec *iov,
                        int iovcnt);
 
-int flux_msg_to_iovec (const flux_msg_t *msg,
+int msg_to_iovec (const flux_msg_t *msg,
                        uint8_t *proto,
                        int proto_len,
                        struct msg_iovec **iovp,
diff --git a/src/common/libzmqutil/msg_zsock.c b/src/common/libzmqutil/msg_zsock.c
index 42e5b9c54..ad71f51fe 100644
--- a/src/common/libzmqutil/msg_zsock.c
+++ b/src/common/libzmqutil/msg_zsock.c
@@ -40,7 +40,7 @@ int zmqutil_msg_send_ex (void *sock, const flux_msg_t *msg, bool nonblock)
         return -1;
     }
 
-    if (flux_msg_to_iovec (msg, proto, PROTO_SIZE, &iov, &iovcnt) < 0)
+    if (msg_to_iovec (msg, proto, PROTO_SIZE, &iov, &iovcnt) < 0)
         goto error;
 
     if (nonblock)
@@ -117,7 +117,7 @@ flux_msg_t *zmqutil_msg_recv (void *sock)
 
     if (!(msg = flux_msg_create (FLUX_MSGTYPE_ANY)))
         goto error;
-    if (flux_iovec_to_msg (msg, iov, iovcnt) < 0)
+    if (iovec_to_msg (msg, iov, iovcnt) < 0)
         goto error;
     rv = msg;
 error:
diff --git a/src/connectors/shmem/Makefile.am b/src/connectors/shmem/Makefile.am
index 82ade229b..1c64c6553 100644
--- a/src/connectors/shmem/Makefile.am
+++ b/src/connectors/shmem/Makefile.am
@@ -22,6 +22,7 @@ shmem_la_LDFLAGS = -module $(san_ld_zdef_flag) \
 shmem_la_LIBADD = \
        $(top_builddir)/src/common/libflux-internal.la \
        $(top_builddir)/src/common/libflux-core.la \
+       $(top_builddir)/src/common/libflux/libflux.la \
        $(top_builddir)/src/common/libzmqutil/libzmqutil.la \
        $(ZMQ_LIBS)

Reply:
I thought of that last night but it felt sort of wrong.  I guess the linker just does magic to resolve and make sure all symbols are good?  I'll give it a shot and see how it turns out.

Reply:
There's plenty of bad libtool juju to contend with, so IMHO when the magic works, go with it. :-)

Reply:
adding `common/libflux/libflux.la` is causing the segfault in the sharness tests again for me.  It commonly comes with `t0009-dmesg.t`.

```
(gdb) bt
#0  0x00002aaaaf010a30 in ?? ()
#1  0x00002aaaaad2938e in fzlistx_delete (self=self@entry=0x2aaab4001240, handle=handle@entry=0x0) at zlistx.c:408
#2  0x00002aaaaad2957a in fzlistx_purge (self=self@entry=0x2aaab4001240) at zlistx.c:465
#3  0x00002aaaaad295ce in fzlistx_destroy (self_p=0x2aaab40096a0) at zlistx.c:131
#4  0x00002aaaaacef75f in flux_msglist_destroy (l=0x2aaab40096a0) at msglist.c:66
#5  0x00002aaaaace92ae in flux_handle_destroy (h=0x2aaab4001170) at handle.c:393
#6  0x000000000040ae72 in module_thread (arg=0x650ef0) at module.c:231
#7  0x00002aaaab16bea5 in start_thread (arg=0x2aaaaefd4700) at pthread_create.c:307
#8  0x00002aaaac2b39fd in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

I'm pretty perplexed by this, as the last function pointer is off into undefined space.  My assumption was that there was a symbol collision during linking.  Or there is some memory corruption going on otherwise not related to this.  But when I look at the output from `nm`, this kinda stuff bothers me:

```
0000000000426e10 T flux_msg_get_seq
0000000000426f40 T flux_msg_get_status
0000000000427b20 T flux_msg_get_string
0000000000427eb0 T flux_msg_get_topic
00000000004268a0 T flux_msg_get_type
0000000000426aa0 T flux_msg_get_userid
                 U flux_msg_handler_addvec
                 U flux_msg_handler_allow_rolemask
                 U flux_msg_handler_create
                 U flux_msg_handler_delvec
                 U flux_msg_handler_destroy
                 U flux_msg_handler_start
                 U flux_msg_handler_stop
0000000000427880 T flux_msg_has_payload
00000000004262b0 T flux_msg_incref
0000000000426a50 T flux_msg_is_noresponse
0000000000426990 T flux_msg_is_private
00000000004269f0 T flux_msg_is_streaming
0000000000427db0 T flux_msg_last_error
0000000000427a90 T flux_msg_pack
0000000000426fc0 T flux_msg_route_clear
```

Some symbols in the `libflux-core.la` shared object are bring brought in and some aren't.

I think the best course of action might be to make a `src/common/libflux/libmessageprivate.la` that `libzmqutil` can ldadd for itself.

Reply:
Hmm, apparently `message.o` was statically linked and `msg_handler.o` will be dynamically linked.  Maybe that is because symbols in `message.o` are referenced by the new convenience library, and `msg_handler.o` symbols are being referenced by the link target?

On the segfault, looks like purge is calling delete with a NULL list entry handle.  The final symbol _would_ be a destructor function pointer, but maybe that pointer is pointing somewhere unresolveable because the list is corrupted.  No idea why rearranging dynamic loading would uncover this but in case it's a real bug, it might be good to run that test under valgrind and see what pops up?  (Happy to help if you want to share your development branch)

Interesting that the stack trace is for a broker module but the test uses the `minimal` personality, which I thought didn't load any modules?

Reply:
So digging into the segfault, I'm quite perplexed.  valgrind didn't catch anything and nothing in gdb looked out of whack.

The fact that gdb says the pointer goes off to nowhere concerned me.  I added a bunch of printfs just to try and see what was going on and eventually came to this, which I've never seen before:

```
flux_msglist_create:66 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaf010b10
flux_handle_create:336 handle=0x2aaab0001190 msglist=0x2aaab00096c0 msglist destructor=0x2aaaaf010b10

<snip>

flux_handle_destroy:394 handle=0x2aaab0001190 msglist=0x2aaab00096c0 msglist destructor=0x2aaaaf010b10
flux_msglist_destroy:74 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaacef740
fzlistx_purge:465 - zlistx destructor = 0x2aaaaf010b10
```

Basically just printing out some pointer values in some create/destroy paths (and pointers to data structures so I match up creates & destroys).  Note the changing value for `msg_destructor` above.  This is the value of the `msg_destructor()` function in `libflux/msglist.c`.   It's not a stored value in some struct.  I don't think the address location of a function should ever change?  The pointer to that function that was stored probably is illegal, thus the segfault.

Then I added more printf debugging and saw this horror:

```
flux_msglist_pollfd:189 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaacef740
flux_msglist_pollevents:198 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaf010b10
flux_msglist_pollevents:198 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaf010b10
flux_msglist_pollevents:198 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaf010b10
flux_msglist_pollevents:198 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaf010b10
flux_msglist_count:182 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaf010b10
flux_msglist_pollevents:198 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaacef740
flux_msglist_pollevents:198 msglist = 0x2aaab00096c0, msg_destructor = 0x2aaaaacef740
```

The address of the `msg_destructor()` function changed over time.

Theory ... could the same symbol have been linked in twice, and there are issues depending on which function is being called?

Reply:
That reproduces for me.

It also makes no sense!  Function addresses are supposed to be constant.  Ug.  I'm going to need to sleep on this one.

Reply:
I don't know internally how a binary "looks up" a symbol when it calls a function, but I'm wondering if the binary linked symbol and the symbol in the shared object are both looked up at times.  Doing another debugging run, the function address of `flux_msglist_pollevents()` also changes.

Reply:
Yeah, no further enlightenment here.:q


Is it possible to inline those iovec functions as a quick solution?

Also, do the PROTO offsets and stuff need to be in that header?  Seems like only PROTO_SIZE needs to be exported?

Reply:
Yeah, the proto export is just for the PROTO_SIZE.

Reply:
Speaking to some DEG folks, there is some scanning stuff that is done internally to find symbols and that scanning can change over time, such as if libraries are used in a certain order / loaded in certain order.  Setting LDDEBUG=all, I noticed this:

```
    161371:	binding file /g/g0/achu/chaos/git/flux-framework/flux-core/src/common/.libs/libflux-core.so.2 [0] to /g/g0/achu/chaos/git/flux-framework/flux-core/src/common/.libs/libflux-core.so.2 [0]: normal symbol `flux_msglist_create'
    161452:	binding file /g/g0/achu/chaos/git/flux-framework/flux-core/src/common/.libs/libflux-core.so.2 [0] to /g/g0/achu/chaos/git/flux-framework/flux-core/src/common/.libs/libflux-core.so.2 [0]: normal symbol `flux_msglist_create'
    161452:	binding file /g/g0/achu/chaos/git/flux-framework/flux-core/src/broker/.libs/lt-flux-broker [0] to /g/g0/achu/chaos/git/flux-framework/flux-core/src/common/.libs/libflux-core.so.2 [0]: normal symbol `flux_msglist_create'
```

I don't know exactly how to read this output, but it does look like `flux_msglist_create()` was found in the .so file and the `lt-flux-broker`.

Reply:
So  I'm trying the `libmessageprivate.la` approach now.  Unfortunately we have to put more stuff into `message_private.h` and `message_private.c` for things to work, but it seems to more correct solution.